pd-client/client.go

```diff
@@ -14,106 +14,401 @@
 
 import (
+	"net"
+	"net/url"
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/juju/errors"
 	"github.com/ngaut/log"
 	"github.com/pingcap/kvproto/pkg/metapb"
-	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/pdpb2"
+	"github.com/pingcap/pd/pkg/apiutil"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
 )
 
 // Client is a PD (Placement Driver) client.
 // It should not be used after calling Close().
 type Client interface {
 	// GetClusterID gets the cluster ID from PD.
-	GetClusterID() uint64
+	GetClusterID(ctx context.Context) uint64
 	// GetTS gets a timestamp from PD.
-	GetTS() (int64, int64, error)
+	GetTS(ctx context.Context) (int64, int64, error)
 	// GetRegion gets a region and its leader Peer from PD by key.
 	// The region may expire after split. Caller is responsible for caching and
 	// taking care of region change.
 	// Also it may return nil if PD finds no Region for the key temporarily,
 	// client should retry later.
-	GetRegion(key []byte) (*metapb.Region, *metapb.Peer, error)
+	GetRegion(ctx context.Context, key []byte) (*metapb.Region, *metapb.Peer, error)
 	// GetStore gets a store from PD by store id.
 	// The store may expire later. Caller is responsible for caching and taking care
 	// of store change.
-	GetStore(storeID uint64) (*metapb.Store, error)
+	GetStore(ctx context.Context, storeID uint64) (*metapb.Store, error)
 	// Close closes the client.
 	Close()
 }
 
+type tsoRequest struct {
+	done     chan error
+	physical int64
+	logical  int64
+}
+
+const (
+	pdTimeout             = 3 * time.Second
+	maxMergeTSORequests   = 10000
+	maxInitClusterRetries = 100
+)
+
+var (
+	// errFailInitClusterID is returned when failed to load clusterID from all supplied PD addresses.
+	errFailInitClusterID = errors.New("[pd] failed to get cluster id")
+	// errClosing is returned when request is canceled when client is closing.
+	errClosing = errors.New("[pd] closing")
+	// errTSOLength is returned when the number of response timestamps is inconsistent with request.
+	errTSOLength = errors.New("[pd] tso length in rpc response is incorrect")
+)
+
 type client struct {
-	worker *rpcWorker
+	urls        []string
+	clusterID   uint64
+	tsoRequests chan *tsoRequest
+
+	connMu struct {
+		sync.RWMutex
+		clientConns map[string]*grpc.ClientConn
+		leader      string
+	}
+
+	checkLeaderCh chan struct{}
+
+	wg     sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
 }
 
 // NewClient creates a PD client.
 func NewClient(pdAddrs []string) (Client, error) {
 	log.Infof("[pd] create pd client with endpoints %v", pdAddrs)
-	worker, err := newRPCWorker(pdAddrs)
-	if err != nil {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &client{
+		urls:          addrsToUrls(pdAddrs),
+		tsoRequests:   make(chan *tsoRequest, maxMergeTSORequests),
+		checkLeaderCh: make(chan struct{}, 1),
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+	c.connMu.clientConns = make(map[string]*grpc.ClientConn)
```

Comment:
I think it's unnecessary. We don't have to connect every pd servers if leader does not change.